### DISCUSSION TITLE
Programmatically updating ImageSource causes image to flicker

### DISCUSSION THREAD

**Discussion by CodedBeard** [dotnet/maui#2560] [Q&A] [answered]
I'm trying to port a WPF app to Maui. The app is pretty simple and just displays a chosen cctv camera feed.

I started by adding an Image control to the main page like so

```xaml
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MauiApp1.MainPage"
             >
    <Image x:Name="Image" WidthRequest="1920" HeightRequest="1080" ></Image>
    
</ContentPage>

```

I then update the ImageSource like so:

```csharp
Device.BeginInvokeOnMainThread(() =>
            {
                Image.Source = ImageSource.FromStream(() => { return new MemoryStream(e.Image); });
            });
```

While this 'works' it results in the image flickering to black every time it is updated, which as the cctv is producing images at 30FPS is completely unuseable.

I found multiple issues reported for xamarin.forms, but all of the suggested fixes don't seem to work in Maui (e.g setting the size of the image).

Is there a better way to update the image so that it doesn't cause it to flicker, or am i just missing another setting somewhere to stop it?

**Comment by mobilewares** [comment, marked as answer]
I had same issue I needed to solve - this is what I did to fix it (via SkiaSharp - nuget package **SkiaSharp.Views.Maui.Controls**).

1) Make your own replacement control - base it on the SkCanvasView from SkiaSharp (MAUI version)
2) Add an eventhandler for the PaintSurface event (so you can custom draw the contents of the control).
3) Write some code to download the image (from HTTP source) into a SKBitmap. This will also allow you to handle things like authenticated access to the IPTV camera.  I have a timer on my host page which calls this every 5 seconds on background thread.
4) Each time the image is updated - invalidate the SK Canvas (ie. this.InvalidateSurface()).  
5) In the PaintSurface event you can then paint the SKBitmap to the canvas (SKCanvas).

This will give you nice clean flicker free drawing (and the ability to manipulate the image if needed too - such as fixing aspect ratio's etc).

**Reply by CodedBeard** [reply]
Thanks for the tip, managed to get it working.
